File: src/day_23/part2.py
from collections import defaultdict
import itertools


lines = open("sample.txt").readlines()
connections = [tuple(line.strip().split("-")) for line in lines]
adjs = defaultdict(set)
networks = {v: [{v}] for v in set(itertools.chain(*connections))}
for v1, v2 in connections:
    adjs[v1].add(v2)
    adjs[v2].add(v1)
    for network in networks[v2]:
        if network < adjs[v1]:
            for v in (new_network := network | {v1}):
                networks[v].append(new_network)
PART_2 = ','.join(sorted(max(itertools.chain.from_iterable(networks.values()), key=len)))
print(PART_2)

File: src/day_23/part2.rs
use std::collections::{HashMap, HashSet};

use itertools::Itertools;

pub fn part2(lines: &[String]) -> String {
  let mut connection_maps = HashMap::new();
  lines.iter().for_each(|l| {
    let parts = l.split("-").collect::<Vec<&str>>();
    connection_maps.entry(parts[0]).or_insert(HashSet::new()).insert(parts[1]);
    connection_maps.entry(parts[1]).or_insert(HashSet::new()).insert(parts[0]);
  });
  let mut connected: Vec<HashSet<String>> = vec![];
  connection_maps.iter().for_each(|(k, first_connections)| {
    first_connections.iter().for_each(|v_2| {
      for c in connected.iter_mut() {
        if c.contains(*v_2) && c.iter().all(|m| first_connections.contains(m.as_str())) {
          c.insert(k.to_string());
        }
      };
      connected.push(HashSet::from([k.to_string(), v_2.to_string()]));
    });
  });
  let sorted = connected.iter().sorted_by(|a, b| b.len().cmp(&a.len())).collect::<Vec<&HashSet<String>>>();
  let sorted_result = sorted[0].iter().sorted().collect::<Vec<&String>>();
  sorted_result.iter().join(",")
}
